man/qatom.1
.\" generated by mkman.py, please do NOT edit!
.TH qatom "1" "Aug 2025" "Gentoo Foundation" "qatom"
.SH NAME
qatom \- split atom strings
.SH SYNOPSIS
.B qatom
\fI[opts] <pkg>\fR
.SH DESCRIPTION
\fIqatom\fR parses strings into atoms and optionally compares them.  The
parsing into atoms results in CATEGORY, PN (package name), PV (package
version), PR (package revision), SLOT and REPO.  Next to these version
qualifiers (e.g.\ >, <, or =) and optional * suffix are extracted.  All
but PN are optional.

Comparison of atoms returns the relationship between two parsed atoms.
\fIqatom\fR does not allow to test conditions.  It is possible to
perform multiple comparisons by simply adding more arguments.
Comparisons are, however, always performed two by two, so there must be
an even count of arguments to the compare function.

The \fIqatom\fR applet is a direct wrapper around the atom parsing
functionalities used by various other applets.  As such this applet
allows easy testing or atom parsing and comparison behaviour.
.SH OPTIONS
.TP
\fB\-F\fR \fI<arg>\fR, \fB\-\-format\fR \fI<arg>\fR
Specify a custom  output  format.  The default format is
.nf
%{CATEGORY} %{PN} %{PV} %[PR] %[SLOT] %[pfx] %[sfx]
.fi
Conversion specifiers start with a \fI%\fR symbol and are followed
by either \fI{\fR or \fI[\fR.  Next is the name of the field to
expand, followed by a matching \fI}\fR or \fI]\fR.  The difference
between \fI{\fR and \fI[\fR is that the latter is only printed if
the field referred is set, while the former prints \fI<unset>\fR in
that case.

The following fields are supported, which are inline with the
variables from \fBebuild\fR(5).
.RS
.IP CATEGORY
The category of the package.
If this field is specified as \fI%[CATEGORY]\fR and it precedes a subsequent
field start symbol, the field will be suffixed with a \fI/\fR.
.IP P
The package name and version without the ebuild revision.
.IP PN
The package name.
.IP PV
The package version without the ebuild revision.
.IP PVR
The package version including ebuild revision, which is also printed
when zero, unlike for \fIPF\fR.
.IP PF
The package name, version and revision when not zero.  Thus, a zero
revision \fI\-r0\fR is not printed.
.IP PR
The ebuild revision, when force printed, outputs \fI\-r0\fR when unset.
.IP SLOT
The package slot, when force printed, outputs \fI\-\fR when unset.
.IP REPO
The package repository.
.IP pfx
The package prefixes, that is version specifiers.
.IP sfx
The package suffices, currently that is just the asterisk.
.RE
.TP
\fB\-c\fR, \fB\-\-compare\fR
Compare two atoms.
.TP
\fB\-s\fR, \fB\-\-scompare\fR
Compare two atoms in the given order data, query.
.TP
\fB\-p\fR, \fB\-\-print\fR
Print reconstructed atom.
.TP
\fB\-l\fR, \fB\-\-lookup\fR
Lookup atom in tree.
.TP
\fB\-\-root\fR \fI<arg>\fR
Set the ROOT env var.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Force all expansions, basically treat all \fI[\fR like \fI{\fR.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Ignored for compatibility with other qapplets.
.TP
\fB\-C\fR, \fB\-\-nocolor\fR
Ignored for compatibility with other qapplets.
.TP
\fB\-\-color\fR
Force color in output.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print this help and exit.
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit.
.SH "ANTISLOT"
A feature present in portage-utils is the so-called antislot, and is
activated by starting the atom with a carrot (^), in place of the
blocker bang (!).  The antislot is similar to the inversing behaviour of
a blocker, but only operates on SLOT and SUBSLOT, and requires SLOT to
be available, e.g. it won't match unset SLOT (NULL).

The antislot is mostly useful with dependencies and is best described
with an example.  Consider a perl upgrade from \fIperl:0/5.28\fR to
\fIperl:0/5.30\fR.  To find a consumer for \fIperl:0/5.28\fR a simple
match can be made, e.g. using \fBqdepends\Rf(1):

.nf
	$ \fIqdepends -Q perl:0/5.28\fR
	virtual/perl-ExtUtils-MakeMaker-7.340.0-r1: ... dev-lang/perl:0/5.28 ...
	...
.fi

However, to query after \fIperl:0/5.30\fR is installed, which packages
are still not rebuilt (e.g. depending on an older perl: 5.28, 5.26,
etc.) one can use the antislot:

.nf
	$ \fIqdepends -Q ^perl:0/5.30\fR
	...
.fi

This will return any package that depends on \fIperl:0\fR not having
subslot \fI5.30\fR.
.P
Obviously this can be tested using \fBqatom\fR using the \fI-c\fR
option:

.nf
	$ \fIqatom -c ^perl:0/5.30 perl-5.28.1:0/5.28\fR
	^perl:0/5.30 == perl-5.28.1:0/5.28
.fi
.SH "REPORTING BUGS"
Please report bugs via http://bugs.gentoo.org/
.br
Product: Gentoo Linux; Component: Current packages
.SH AUTHORS
.nf
Ned Ludd <solar@gentoo.org>
Mike Frysinger <vapier@gentoo.org>
Fabian Groffen <grobian@gentoo.org>
.fi
.SH "SEE ALSO"
.BR q (1),
.BR qcheck (1),
.BR qdepends (1),
.BR qfile (1),
.BR qgrep (1),
.BR qkeyword (1),
.BR qlist (1),
.BR qlop (1),
.BR qmanifest (1),
.BR qmerge (1),
.BR qpkg (1),
.BR qsearch (1),
.BR qsize (1),
.BR qtbz2 (1),
.BR qtegrity (1),
.BR quse (1),
.BR qwhich (1),
.BR qxpak (1)


qatom.c
/*
 * Copyright 2005-2019 Gentoo Foundation
 * Distributed under the terms of the GNU General Public License v2
 *
 * Copyright 2005-2008 Ned Ludd        - <solar@gentoo.org>
 * Copyright 2005-2014 Mike Frysinger  - <vapier@gentoo.org>
 * Copyright 2018-     Fabian Groffen  - <grobian@gentoo.org>
 */

#include "main.h"
#include "applets.h"

#include "atom.h"
#include "tree.h"

#define QATOM_FORMAT "%{CATEGORY} %{PN} %{PV} %[PR] %[SLOT] %[pfx] %[sfx]"

#define QATOM_FLAGS "F:cspl" COMMON_FLAGS
static struct option const qatom_long_opts[] = {
	{"format",     a_argument, NULL, 'F'},
	{"compare",   no_argument, NULL, 'c'},
	{"scompare",  no_argument, NULL, 's'},
	{"print",     no_argument, NULL, 'p'},
	{"lookup",    no_argument, NULL, 'l'},
	COMMON_LONG_OPTS
};
static const char * const qatom_opts_help[] = {
	"Custom output format (default: " QATOM_FORMAT ")",
	"Compare two atoms",
	"Compare two atoms in the given order data, query",
	"Print reconstructed atom",
	"Lookup atom in tree",
	COMMON_OPTS_HELP
};
#define qatom_usage(ret) usage(ret, QATOM_FLAGS, qatom_long_opts, qatom_opts_help, NULL, lookup_applet_idx("qatom"))

int qatom_main(int argc, char **argv)
{
	enum qatom_atom {
		_EXPLODE = 0,
		_COMPARE,
		_SCOMPARE,
		_PRINT,
		_LOOKUP
	} action = _EXPLODE;
	const char *format = QATOM_FORMAT;
	depend_atom *atom;
	depend_atom *atomc;
	tree_ctx *tree = NULL;
	int i;

	while ((i = GETOPT_LONG(QATOM, qatom, "")) != -1) {
		switch (i) {
		case 'F': format = optarg;    break;
		case 'c': action = _COMPARE;  break;
		case 's': action = _SCOMPARE; break;
		case 'p': action = _PRINT;    break;
		case 'l': action = _LOOKUP;   break;
		COMMON_GETOPTS_CASES(qatom)
		}
	}

	if (argc == optind)
		qatom_usage(EXIT_FAILURE);

	if ((action == _COMPARE || action == _SCOMPARE) && (argc - optind) % 2)
		err("compare needs even number of arguments");

	if (action == _LOOKUP) {
		tree = tree_open(portroot, main_overlay);
		if (tree == NULL)
			err("failed to open tree");
	}

	for (i = optind; i < argc; i++) {
		atom = atom_explode(argv[i]);
		if (atom == NULL) {
			warnf("invalid atom: %s\n", argv[i]);
			continue;
		}

		switch (action) {
		case _COMPARE:
		case _SCOMPARE: {
			int r;

			i++;
			atomc = atom_explode(argv[i]);
			if (atomc == NULL) {
				warnf("invalid atom: %s\n", argv[i]);
				break;
			}

			if (action == _SCOMPARE ||
					atomc->blocker != ATOM_BL_NONE ||
					atomc->pfx_op != ATOM_OP_NONE ||
					atomc->sfx_op != ATOM_OP_NONE ||
					(atomc->CATEGORY == NULL &&
					 atom->blocker == ATOM_BL_NONE &&
					 atom->pfx_op == ATOM_OP_NONE &&
					 atom->sfx_op == ATOM_OP_NONE))
			{
				r = atom_compare(atom, atomc);
			} else {
				r = atom_compare(atomc, atom);
				switch (r) {
					case NEWER:     r = OLDER;     break;
					case OLDER:     r = NEWER;     break;
				}
			}

			printf("%s %s ", atom_to_string(atom), booga[r]);
			printf("%s\n", atom_to_string(atomc));
			atom_implode(atomc);
			break;
		}
		case _EXPLODE:
			printf("%s\n", atom_format(format, atom));
			break;
		case _PRINT:
			printf("%s\n", atom_to_string(atom));
			break;
		case _LOOKUP:
			{
				tree_match_ctx *pkg = tree_match_atom(tree, atom,
						TREE_MATCH_DEFAULT);
				if (pkg != NULL) {
					tree_match_ctx *w;

					for (w = pkg; w != NULL; w = w->next) {
						if (!quiet)
							printf("%s: ", atom_to_string(atom));
						printf("%s\n", atom_format(format, w->atom));
					}

					tree_match_close(pkg);
				}
			}
		}

		atom_implode(atom);
	}

	if (action == _LOOKUP)
		tree_close(tree);

	return EXIT_SUCCESS;
}
